File: bachetsgame.cpp
#include <iostream>
#include <vector>
#include <array>

using namespace std;

int main() {
    int n;
    int m;
    while (cin >> n) {
        cin >> m;
        int x;
        vector<int> moves;
        for (int i=0; i < m; i++){
            cin >> x;
            moves.push_back(x);
        }

        array<int, 1000001> dp;
        for (int i=0; i<n+1; i++){
            dp[i] = 0;
            for (int j=0; j<m; j++){
                if (i - moves[j] == 0){
                    dp[i] = 1;
                    break;
                } else {
                    bool possible;
                    possible = i - moves[j] > 0;
                    if (possible) {
                        if (dp[i-moves[j]] != 1){
                            dp[i] = 1;
                            break;
                        }
                    }
                }
            }
        }
        int w = dp[n];
        if (w == 1) {
            cout << "Stan wins" << endl;
        } else {
            cout << "Ollie wins" << endl;
        }

    }
}

File: bachetsgame.py
import sys


for line in sys.stdin:
    n, m, moves = line.split(' ', 2)
    n = int(n)
    m = int(m)
    moves = list(map(int, moves.split()))
    dp = [[-1 for x in range(m+1)] for y in range(n+1)]
    for i in range(n+1):
        for j in range(m):
            if i - moves[j] == 0:
                dp[i][j] = 1
                dp[i][-1] = 1
            else:
                possible = i - moves[j] > 0
                if possible:
                    outcome = dp[i-moves[j]]
                    # if opponent can win off this move, then
                    # mark it 0 (opponent wins)
                    if outcome[-1] == 1:
                        dp[i][j] = 0 
                    else:
                        dp[i][j] = 1 
                        dp[i][-1] = 1

    w = dp[i][-1] 
    if w == 1:
        print('Stan wins')
    else:
        print('Ollie wins')
